## Student Dataset

In [1]:
import numpy as np
import pandas as pd

The 'ClassificationDataSetforStudents.xls' file was converted to two csv files: 'traindata.csv' and 'testdata.csv'.

In [2]:
train = pd.read_csv('traindata.csv', header=None)
test = pd.read_csv('testdata.csv', header=None)

In [3]:
train_X = train.drop(train.columns[23], axis=1)
train_y = train.iloc[:,23]

In [4]:
train_X

0   1   2   3   4   5   6   7   8     9   ...      13      14      15  \
0       5   1   2   2  -1  -1  -1   2   0  5945  ...    3897    6978    4081   
1       5   1   1  -1  -1  -2  -1  -1  -1  5945  ...       0     150     150   
2       3   1   1   0   0   0   0   0   0  7175  ...  115177  116226  139737   
3       3   3   2   0   0   2   0   0  -1  5125  ...   39909   18334    7382   
4       5   1   1   0  -1  -1  -1  -1  -1  7790  ...    1156     316     316   
...    ..  ..  ..  ..  ..  ..  ..  ..  ..   ...  ...     ...     ...     ...   
25995   3   3   1   0   0   0   0   0   0  7995  ...  208365   88004   31237   
25996   3   3   2  -1  -1  -1  -1   0   0  8815  ...    3502    8979    5190   
25997   3   2   2   4   3   2  -1   0   0  7585  ...    2758   20878   20582   
25998   3   3   1   1  -1   0   0   0  -1  8405  ...   76304   52774   11855   
25999   3   2   1   0   0   0   0   0   0  9430  ...   49764   36535   32428   

           16     17     18     19     20     21     22  
0        3686   1533   3897   4081      0    155     96  
1         989      0      0    150    150    989      0  
2      187728   8500   6000   4000  25000  50000  70000  
3       19801   7504     29  10004      6  19824  18129  
4         316    316   1156    316    316    316    316  
...       ...    ...    ...    ...    ...    ...    ...  
25995   15980   8500  20000   5003   3047   5000   1000  
25996       0   1837   3526   8998    129      0      0  
25997   19357      0      0  22000   4200   2000   3100  
25998   48944  85900   3409   1178   1926  52964   1804  
25999   15313   2078   1800   1430   1000   1000   1000  

[26000 rows x 23 columns]

In [5]:
train_y

0        0
1        1
2        1
3        0
4        0
        ..
25995    0
25996    0
25997    1
25998    1
25999    1
Name: 23, Length: 26000, dtype: int64

Comparison of range

In [6]:
rnge = pd.concat([train_X.min(),test.min(),train_X.max(),test.max(),
                    train_X.max()-train_X.min(),test.max()-test.min()], axis=1)
rnge.columns = ['train_min','test_min','train_max','test_max','train_range', 'test_range']
rnge

train_min  test_min  train_max  test_max  train_range  test_range
0           3         3          5         5            2           2
1           0         0          6         6            6           6
2           0         0          3         3            3           3
3          -2        -2          8         8           10          10
4          -2        -2          8         7           10           9
5          -2        -2          8         7           10           9
6          -2        -2          8         7           10           9
7          -2        -2          8         7           10           9
8          -2        -2          8         8           10          10
9        4305      4305      16195     15375        11890       11070
10      10000     10000     800000   1000000       790000      990000
11    -165580    -14386     746814    964511       912394      978897
12     -69777    -24704     743970    983931       813747     1008635
13    -157264    -17706    1664089    578971      1821353      596677
14     -81334   -170000     706864    891586       788198     1061586
15     -81334    -28335     823540    927171       904874      955506
16    -209051   -339603     568638    961664       777689     1301267
17          0         0     873552    873552       873552      873552
18          0         0    1684259   1215471      1684259     1215471
19          0         0     889043    896040       889043      896040
20          0         0     621000    621000       621000      621000
21          0         0     426529    332000       426529      332000
22          0         0     527143    528666       527143      528666

Comparison of mean

In [7]:
mean = pd.concat([train_X.mean(), test.mean()], axis=1)
mean.columns = ['train_mean', 'test_mean']
mean

train_mean    test_mean
0        4.216000       4.2956
1        1.858462       1.8530
2        1.549077       1.5410
3       -0.018308      -0.0424
4       -0.131308      -0.1814
5       -0.166654      -0.1910
6       -0.216077      -0.2738
7       -0.265577      -0.3016
8       -0.289462      -0.3302
9     7281.844423    7378.6880
10  167693.833846  169096.0000
11   51499.240423   49257.0590
12   49416.018692   47296.0514
13   47346.507962   44721.1548
14   43724.758615   40650.7026
15   40532.023769   38680.8152
16   39097.610000   37196.3922
17    5671.374269    5775.4620
18    5977.154846    5841.9232
19    5331.572500    5031.0452
20    4845.287077    4713.3838
21    4792.553846    4797.2950
22    5194.856962    5538.4670

Comparison of variance

In [8]:
variance = pd.concat([train_X.var(), test.var()], axis=1)
variance.columns = ['train_variance', 'test_variance']
variance

train_variance  test_variance
0     9.533807e-01   9.128032e-01
1     6.306065e-01   6.171144e-01
2     2.720635e-01   2.799750e-01
3     1.264483e+00   1.254853e+00
4     1.429352e+00   1.455985e+00
5     1.419858e+00   1.508421e+00
6     1.370133e+00   1.362306e+00
7     1.279403e+00   1.294496e+00
8     1.315186e+00   1.351038e+00
9     3.565051e+06   3.673012e+06
10    1.679028e+10   1.706948e+10
11    5.400237e+09   5.554079e+09
12    5.039201e+09   5.156179e+09
13    4.835932e+09   4.701978e+09
14    4.137280e+09   4.154055e+09
15    3.687565e+09   3.748447e+09
16    3.514823e+09   3.692784e+09
17    2.814853e+08   3.714248e+08
18    5.693446e+08   5.812533e+08
19    3.028725e+08   5.287263e+08
20    2.487522e+08   2.751949e+08
21    2.321545e+08   2.357819e+08
22    3.009146e+08   4.468652e+08

In [9]:
train_y.describe()

count    26000.000000
mean         0.220923
std          0.414877
min          0.000000
25%          0.000000
50%          0.000000
75%          0.000000
max          1.000000
Name: 23, dtype: float64

In [10]:
train_y.value_counts()

0    20256
1     5744
Name: 23, dtype: int64

In [11]:
train_y.value_counts()[0]/len(train_y)

0.7790769230769231

In [12]:
train_y.value_counts()[1]/len(train_y)

0.22092307692307692

The classes are not balanced as we can see above, since class 0 consists of 77.9% of the training data while class 1 consists of only 22.1% of the training data.

## Decision Tree Classifier

In [13]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import StratifiedKFold
from sklearn.model_selection import cross_val_score
from sklearn import tree
from matplotlib import pyplot as plt

In [14]:
clf = DecisionTreeClassifier(random_state=0)
clf.fit(train_X, train_y)

DecisionTreeClassifier(random_state=0)

In [15]:
metrics = ['precision', 'recall', 'accuracy']
for m in metrics:
    scores = cross_val_score(clf, train_X, train_y, scoring=m, cv=10)
    print(m, ": ", scores)
    print("average", m, ": ", scores.mean())

precision :  [0.35569423 0.396875   0.36080871 0.37952756 0.37293233 0.4147541
 0.42882883 0.40353982 0.42831541 0.38897397]
average precision :  0.393024995552109
recall :  [0.39721254 0.44250871 0.40418118 0.41986063 0.43205575 0.44076655
 0.41391304 0.39652174 0.41565217 0.44173913]
average recall :  0.42044114528101806
accuracy :  [0.70807692 0.72846154 0.71038462 0.72038462 0.71423077 0.73923077
 0.74846154 0.73692308 0.74807692 0.72307692]
average accuracy :  0.7277307692307692


## AdaBoost

In [18]:
from sklearn.ensemble import AdaBoostClassifier

In [21]:
boosted_clf = AdaBoostClassifier()
boosted_clf.fit(train_X, train_y)

AdaBoostClassifier()

In [22]:
metrics = ['precision', 'recall', 'accuracy']
for m in metrics:
    scores = cross_val_score(boosted_clf, train_X, train_y, scoring=m, cv=10)
    print(m, ": ", scores)
    print("average", m, ": ", scores.mean())

precision :  [0.65454545 0.65037594 0.62962963 0.67028986 0.67619048 0.7284345
 0.71052632 0.73839662 0.74297189 0.72614108]
average precision :  0.6927501766730404
recall :  [0.31358885 0.30139373 0.29616725 0.32229965 0.37108014 0.39721254
 0.32869565 0.30434783 0.32173913 0.30434783]
average recall :  0.32608725950613543
accuracy :  [0.81192308 0.81       0.80615385 0.81538462 0.82192308 0.83423077
 0.82192308 0.82230769 0.82538462 0.82076923]
average accuracy :  0.8190000000000002


In [25]:
# different learning rates
lr = [0.001, 0.01, 0.1]
for r in lr:
    boosted_clf = AdaBoostClassifier(learning_rate=r)
    print("=================================================")
    print("learning rate", r)
    for m in metrics:
        scores = cross_val_score(boosted_clf, train_X, train_y, scoring=m, cv=10)
        print(m, ": ", scores)
        print("average", m, ": ", scores.mean())

learning rate 0.001
precision :  [0.64130435 0.66788321 0.62921348 0.69402985 0.69736842 0.75786164
 0.71691176 0.75819672 0.74596774 0.7265625 ]
average precision :  0.7035299677622854
recall :  [0.30836237 0.31881533 0.29268293 0.32404181 0.36933798 0.41986063
 0.33913043 0.32173913 0.32173913 0.32347826]
average recall :  0.33391880018179065
accuracy :  [0.80923077 0.81461538 0.80576923 0.81923077 0.82538462 0.84230769
 0.82423077 0.82730769 0.82576923 0.82346154]
average accuracy :  0.8217307692307692
learning rate 0.01
precision :  [0.64130435 0.66788321 0.62921348 0.69402985 0.69736842 0.75786164
 0.71691176 0.75819672 0.74596774 0.7265625 ]
average precision :  0.7035299677622854
recall :  [0.30836237 0.31881533 0.29268293 0.32404181 0.36933798 0.41986063
 0.33913043 0.32173913 0.32173913 0.32347826]
average recall :  0.33391880018179065
accuracy :  [0.80923077 0.81461538 0.80576923 0.81923077 0.82538462 0.84230769
 0.82423077 0.82730769 0.82576923 0.82346154]
average accuracy :

Changes in performance after using AdaBoost, with best accuracy obtained using learning rate 0.01:

In [45]:
print((0.7035299677622854-0.393024995552109)/0.393024995552109*100, "percent increase in precision")
print((0.33391880018179065-0.42044114528101806)/0.42044114528101806*100, "percent increase in recall")
print((0.8217307692307692-0.7277307692307692)/0.7277307692307692*100, "percent increase in accuracy")

79.00387398363527 percent increase in precision
-20.578943348039083 percent increase in recall
12.916864859151206 percent increase in accuracy


## Gradient Boosting Classifier

In [28]:
from sklearn.ensemble import HistGradientBoostingClassifier

In [36]:
hgbc = HistGradientBoostingClassifier(categorical_features=[0,1,2], learning_rate=0.05)
hgbc.fit(train_X, train_y)
scores = cross_val_score(hgbc, train_X, train_y, scoring='accuracy', cv=10)
print("accuracy:", scores)
print("average accuracy:", scores.mean())

accuracy: [0.81115385 0.81961538 0.80846154 0.81653846 0.82192308 0.84538462
 0.83538462 0.82807692 0.82769231 0.82423077]
average accuracy: 0.8238461538461539


In [37]:
out = hgbc.predict(test)

In [38]:
out

array([1, 0, 0, ..., 0, 1, 0], dtype=int64)

In [41]:
np.savetxt("output4.txt", out, fmt="%d")